File: apps/api/app/schemas/platform_inventory.py
"""Multi-platform (Amazon/Flipkart/Blinkit/Meesho/Manual + Shopify)
inventory schemas. All quantities are in BOXES, same unit as
`app.schemas.inventory` (see that module's docstring) — a platform-stock
number and a Shopify `available_boxes` number are always directly
comparable, never packets-vs-boxes mismatched.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.enums import PlatformStockMovementType, ProductMarketplaceMovementType
from app.models.platform_inventory import InventoryPlatform


class PlatformStockMovementCreateRequest(BaseModel):
    """Record ONE manual platform stock movement — staff enters only the
    quantity being added or deducted, never the resulting total (mirrors
    `InventoryAdjustmentRequest`'s existing additive-only contract
    exactly; see `PlatformInventoryService.record_movement`). `reason` is
    optional here (unlike the existing Shopify adjustment's mandatory
    reason) — a deliberate difference, not an oversight.
    """

    platform: str = Field(max_length=50)
    movement_type: Literal["stock_added", "stock_deducted"]
    quantity: int = Field(gt=0, description="Quantity being added or deducted, in boxes.")
    reason: str | None = Field(default=None, max_length=255)
    stock_date: date | None = Field(
        default=None,
        description="IST business date this movement belongs to. Defaults to today (IST).",
    )

    @field_validator("platform")
    @classmethod
    def _validate_platform(cls, value: str) -> str:
        if value not in InventoryPlatform.ALL:
            allowed = ", ".join(InventoryPlatform.ALL)
            raise ValueError(f"Unknown platform {value!r}. Must be one of: {allowed}.")
        return value


class PlatformStockMovementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    product_variant_id: uuid.UUID
    platform: str
    platform_label: str
    movement_type: PlatformStockMovementType
    quantity_delta: int
    quantity_after: int
    stock_date: date
    reason: str | None
    actor_user_id: uuid.UUID | None
    actor_label: str
    created_at: datetime


class UnifiedStockMovementResponse(BaseModel):
    """One row of the merged "Platform Stock Movement History" (this
    product variant's `PlatformStockMovement` rows interleaved with its
    existing Shopify `InventoryMovement` rows, sorted by time) — the
    existing standalone Shopify movement-history table/endpoint is
    completely untouched; this is an ADDITIONAL, read-only merged view.
    """

    id: uuid.UUID
    platform: str
    platform_label: str
    movement_type: str
    quantity_delta: int
    quantity_after: int
    stock_date: date
    reason: str | None
    actor_label: str
    created_at: datetime


class PlatformStockSummaryRow(BaseModel):
    """One platform's row in the "Marketplace Stock" table for one
    product variant, on the selected `stock_date`.

    `is_automatic=True` (Shopify): `current_stock` is the true LIVE value
    (`ProductVariant.available_quantity`) only when `stock_date` is
    TODAY (IST) — always exact, never reconstructed, even if some
    non-ledger process ever touched it. For a PAST `stock_date`,
    `current_stock`/`opening_stock` are reconstructed from the existing
    `InventoryMovement` ledger: each movement already records
    `quantity_after` (its resulting running balance), so the latest
    movement at/before a cutoff IS the balance at that instant — the
    exact same technique `PlatformStockMovementRepository.get_latest_
    as_of` already uses for the manual marketplace ledger, just applied
    to Shopify's own existing ledger instead of a second one.
    `stock_added`/`stock_deducted` are always genuinely date-scoped
    movement totals for both Shopify and every manual platform.

    `current_stock`/`opening_stock` are `None` (never a guessed 0) when
    no `InventoryMovement` row exists before the relevant cutoff — the
    ledger genuinely cannot reconstruct that date's Shopify balance (it
    may predate the variant's first-ever movement); the frontend must
    render this as "historical data not available", never as zero.

    `is_automatic=False` (Amazon/Flipkart/Blinkit/Meesho/Manual):
    `current_stock` is the true closing balance AS OF `stock_date`
    (`PlatformStockMovementRepository.get_latest_as_of`), defaulting to
    `0` (never `None`) when no movement has ever been recorded — unlike
    Shopify, "never told about any stock yet" legitimately means 0 for a
    platform whose only source of truth IS this ledger.
    """

    platform: str
    platform_label: str
    is_automatic: bool
    opening_stock: int | None
    stock_added: int
    stock_deducted: int
    current_stock: int | None
    last_updated: datetime | None


class ProductPlatformStockResponse(BaseModel):
    """Marketplace Stock table: ONE row per platform for the whole
    PRODUCT (never one per SKU) -- Shopify's row is summed across every
    real underlying `ProductVariant` (server-side, same null-propagation
    rule `PlatformStockSummaryRow` already documents); every manual
    platform's row comes directly from `ProductMarketplaceMovement`,
    which is already product-scoped and needs no SKU-level summation at
    all.
    """

    product_id: uuid.UUID
    product_title: str
    stock_date: date
    platforms: list[PlatformStockSummaryRow]


class ProductMarketplaceMovementCreateRequest(BaseModel):
    """Record ONE product-level manual marketplace movement (Add Stock /
    Record Sale / RTO) -- NO SKU is selected or implied. The business
    user enters a bare packet quantity for the whole product on this
    platform; see `PlatformInventoryService.record_product_movement` for
    the packet->outer conversion and the uniform-pack-size safety check
    that can reject this request (422) rather than guess an allocation.
    """

    platform: str = Field(max_length=50)
    movement_type: Literal["stock_added", "sale", "rto"]
    quantity_packets: int = Field(gt=0, description="Quantity in packets, as entered by staff.")
    reason: str | None = Field(default=None, max_length=255)
    stock_date: date | None = Field(
        default=None,
        description="IST business date this movement belongs to. Defaults to today (IST).",
    )

    @field_validator("platform")
    @classmethod
    def _validate_platform(cls, value: str) -> str:
        if value not in InventoryPlatform.ALL:
            allowed = ", ".join(InventoryPlatform.ALL)
            raise ValueError(f"Unknown platform {value!r}. Must be one of: {allowed}.")
        return value


class ProductMarketplaceMovementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    product_id: uuid.UUID
    platform: str
    platform_label: str
    movement_type: ProductMarketplaceMovementType
    quantity_packets: int
    quantity_delta: int
    quantity_after: int
    stock_date: date
    reason: str | None
    actor_user_id: uuid.UUID | None
    actor_label: str
    created_at: datetime


class ShipmentTransitSummaryRow(BaseModel):
    """One variant's shipment-status breakdown, derived entirely from the
    existing `InventoryMovement` (DISPATCH rows already link
    `product_variant_id` + `shipment_id`) joined to `Shipment.
    current_status` — never a fabricated/estimated count.

    `in_transit`/`out_for_delivery`/`rto` are LIVE current snapshot
    counts, not scoped to `stock_date` — a shipment's status has no
    historical day-by-day record in this codebase, so filtering these by
    a past date would silently show today's numbers under a stale label,
    which Requirement 10 explicitly forbids. Only `delivered_on_date` is
    genuinely date-scoped (by `Shipment.actual_delivery_date`).
    """

    product_variant_id: uuid.UUID
    sku: str
    in_transit: int
    out_for_delivery: int
    delivered_on_date: int
    rto: int


class ProductShipmentSummaryResponse(BaseModel):
    product_id: uuid.UUID
    stock_date: date
    in_transit: int
    out_for_delivery: int
    delivered_on_date: int
    rto: int
    variants: list[ShipmentTransitSummaryRow]

File: apps/web/types/platform-inventory.ts
/** Mirrors apps/api/app/schemas/platform_inventory.py and
 * apps/api/app/models/platform_inventory.py::InventoryPlatform.
 * All quantities are in BOXES, same unit as types/inventory.ts.
 */

export type ManualPlatform = "amazon" | "flipkart" | "blinkit" | "meesho" | "manual_other"

/** Every value the "Add Stock" / "Record Sale" dialog's Platform field
 * can submit — deliberately a plain string union mirroring the backend's
 * `InventoryPlatform.ALL` (never a hardcoded switch scattered through the
 * app) so adding a platform later is a one-line change here, not a
 * rewrite.
 */
export const MANUAL_PLATFORMS: ManualPlatform[] = [
  "amazon",
  "flipkart",
  "blinkit",
  "meesho",
  "manual_other",
]

export const PLATFORM_LABELS: Record<ManualPlatform | "shopify", string> = {
  shopify: "Shopify",
  amazon: "Amazon",
  flipkart: "Flipkart",
  blinkit: "Blinkit",
  meesho: "Meesho",
  manual_other: "Manual / Other",
}

export type PlatformStockMovementType = "stock_added" | "stock_deducted"

export interface PlatformStockMovementCreateInput {
  platform: ManualPlatform
  movement_type: PlatformStockMovementType
  quantity: number
  reason?: string
  /** `YYYY-MM-DD` (IST). Omit to default to today (IST) on the backend. */
  stock_date?: string
}

export interface PlatformStockMovement {
  id: string
  product_variant_id: string
  platform: string
  platform_label: string
  movement_type: PlatformStockMovementType
  quantity_delta: number
  quantity_after: number
  stock_date: string
  reason: string | null
  actor_user_id: string | null
  actor_label: string
  created_at: string
}

/** One row of the merged "Platform Stock Movement History" table — this
 * variant's manual platform movements interleaved with its existing
 * Shopify movements (`dispatch`/`rto_restock`/`manual_adjustment`/
 * `initial_stock`), sorted by time.
 */
export interface UnifiedStockMovement {
  id: string
  platform: string
  platform_label: string
  movement_type: string
  quantity_delta: number
  quantity_after: number
  stock_date: string
  reason: string | null
  actor_label: string
  created_at: string
}

/** One platform's row in the "Marketplace Stock" table, for one product
 * variant, on the selected stock date.
 *
 * `is_automatic=true` (Shopify): `current_stock` is the true LIVE value
 * only when `stock_date` is TODAY. For a PAST date, `current_stock`/
 * `opening_stock` are reconstructed from Shopify's own existing movement
 * ledger (never fabricated) and are `null` — never `0` — when that
 * ledger has no row before the relevant cutoff (it may predate the
 * variant's first-ever movement). Render `null` as "historical data not
 * available" (e.g. "—"), never as zero — a `null` reading must never be
 * mistaken for "no stock".
 *
 * `is_automatic=false` (manual platforms): `current_stock` defaults to
 * `0` (never `null`) when nothing has ever been recorded — unlike
 * Shopify, this ledger IS the only source of truth for a manual
 * platform, so "nothing recorded" legitimately means 0.
 */
export interface PlatformStockSummaryRow {
  platform: string
  platform_label: string
  is_automatic: boolean
  opening_stock: number | null
  stock_added: number
  stock_deducted: number
  current_stock: number | null
  last_updated: string | null
}

/** ONE OMS-visible variant's own Marketplace Stock (a product with two or
 * more CatalogVariants -- Herbal Masala's Gold/Red/Blue). Its Shopify row
 * is summed across only THAT variant's SKUs; its manual rows come from
 * movements scoped to it, so Amazon Gold never mixes with Amazon Red. The
 * 60/120/180 SKUs never get a table of their own.
 */
export interface VariantMarketplaceStock {
  catalog_variant_id: string
  name: string
  display_order: number
  platforms: PlatformStockSummaryRow[]
  /** Effective packets sold by THIS variant this IST calendar month
   * (Sale only; a reversed Sale is excluded).
   */
  sold_this_month_packets: number
}

/** Marketplace Stock, in one of two shapes:
 *  - `scope: "product"` -- ONE table for the product in `platforms`
 *    (Shopify already summed across SKUs server-side).
 *  - `scope: "catalog_variant"` -- one table PER OMS-visible variant in
 *    `variants` (`platforms` is empty); never combined across variants.
 */
export interface ProductPlatformStock {
  product_id: string
  product_title: string
  stock_date: string
  scope: "product" | "catalog_variant"
  platforms: PlatformStockSummaryRow[]
  variants: VariantMarketplaceStock[]
  /** Packets sold across every manual platform in the CURRENT IST
   * calendar month for a `scope: "product"` product -- effective Sale
   * movements only, computed by the backend from real records (never
   * RTO, never Shopify's automatic movements). 0 for a variant-scoped
   * product, which reports it per variant instead.
   */
  sold_this_month_packets: number
}

/** A marketplace movement's type. `stock_added` is legacy only (no UI or
 * API path creates it any more); `reversal` is a compensating row written
 * by Undo / Edit (never entered directly).
 */
export type ProductMarketplaceMovementType = "stock_added" | "sale" | "rto" | "reversal"

/** What the Record Sale / RTO dialogs may submit -- no SKU. `quantity_packets`
 * is what the business user actually typed; the backend converts it to
 * outers using the in-scope SKUs' pack_size/packets_per_box (422 if those
 * aren't uniform) and applies the same effect to the OMS total stock in
 * one transaction.
 */
export type MarketplaceOperation = "sale" | "rto"

export interface ProductMarketplaceMovementCreateInput {
  platform: ManualPlatform
  movement_type: MarketplaceOperation
  quantity_packets: number
  /** ONLY for a variant-scoped product (Herbal): the OMS-visible variant
   * (Gold/Red/Blue) the movement belongs to -- never an underlying SKU.
   */
  catalog_variant_id?: string
  reason?: string
  /** `YYYY-MM-DD` (IST). Omit to default to today (IST) on the backend. */
  stock_date?: string
}

/** Edit appends a reversal + a replacement; Undo appends a reversal. The
 * original row is never changed. A reason is required for the audit trail.
 */
export interface ProductMarketplaceMovementEditInput {
  quantity_packets: number
  reason: string
}

export interface ProductMarketplaceMovementUndoInput {
  reason: string
}

/** Effective state derived from the reversal/replacement links:
 *  active -- in force; undone -- reversed; edited -- reversed and
 *  replaced; reversal -- a compensating row.
 */
export type MarketplaceMovementStatus = "active" | "undone" | "edited" | "reversal"

/** One row of the Marketplace history -- Sale / RTO / Reversal, each its
 * own event (never merged or netted).
 */
export interface ProductMarketplaceMovement {
  id: string
  product_id: string
  catalog_variant_id: string | null
  platform: string
  platform_label: string
  movement_type: ProductMarketplaceMovementType
  quantity_packets: number
  quantity_delta: number
  quantity_after: number
  stock_date: string
  reason: string | null
  actor_user_id: string | null
  actor_label: string
  created_at: string
  reverses_movement_id: string | null
  replaces_movement_id: string | null
  /** On a replacement row: the quantity it replaced ("edited from 20"). */
  edited_from_packets: number | null
  status: MarketplaceMovementStatus
  can_edit: boolean
  can_undo: boolean
}

/** One variant's shipment-status breakdown — `in_transit`/
 * `out_for_delivery`/`rto` are LIVE current counts (shipment status has
 * no historical day-by-day record), never scoped to the selected date;
 * only `delivered_on_date` is genuinely date-scoped. Never add
 * `in_transit` into `current_stock` — it's stock already dispatched, not
 * available marketplace stock (Requirement 16).
 */
export interface ShipmentTransitSummaryRow {
  product_variant_id: string
  sku: string
  in_transit: number
  out_for_delivery: number
  delivered_on_date: number
  rto: number
}

export interface ProductShipmentSummary {
  product_id: string
  stock_date: string
  in_transit: number
  out_for_delivery: number
  delivered_on_date: number
  rto: number
  variants: ShipmentTransitSummaryRow[]
}
